### PR TITLE
07 security

### PR DESCRIPTION
**Pull request by anastasiia-fi**

### CODE REVIEW

File: src/main/java/com/example/bookstore/dao/UserRepository.java
Comment (yevhenkukharchuk) [inline src/main/java/com/example/bookstore/dao/UserRepository.java:1]: you should add also RoleRepository

**Review by yevhenkukharchuk — CHANGES_REQUESTED** [review]
Big job! But you should continue for finalize this solution.

File: src/main/java/com/example/bookstore/config/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.config.http.SessionCreationPolicy;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
+
+@RequiredArgsConstructor
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/config/SecurityConfig.java:20]: Add annotation EnableMethodSecurity 
Did you test your solution?

File: src/main/java/com/example/bookstore/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers("/auth/**")
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/config/SecurityConfig.java:38]: you also should add 
```
"/swagger-ui/**", 
"/errors"
```

File: src/main/java/com/example/bookstore/controller/AuthController.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            throws RegistrationException {
+        return userService.register(request);
+    }
+
+    @Operation(summary = "Logging in", description = "Logging in for a registered user")
+    @PostMapping("/login")
+    public UserLoginResponseDto login(@RequestBody @Valid UserLoginRequestDto request) {
+        return authenticationService.authenticate(request);
+    }
+
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/controller/AuthController.java:39]:
```suggestion
```

File: src/main/java/com/example/bookstore/dao/RoleRepository.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package com.example.bookstore.dao;
+
+import com.example.bookstore.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/dao/RoleRepository.java:6]: Repository annotation missed

File: src/main/java/com/example/bookstore/dto/user/UserDto.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.example.bookstore.dto.user;
+
+public record UserDto(
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/dto/user/UserDto.java:3]: do you use this Dto?

File: src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package com.example.bookstore.dto.user;
+
+import com.example.bookstore.validator.FieldMatch;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Size;
+
+@FieldMatch
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java:8]: Let’s add parameters ;)
(message = "Password don`t match",
        field = "password",
        fieldMatch = "verifyPassword")

File: src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        @Size(min = 6, max = 100)
+        String password,
+
+        @NotBlank
+        @Size(min = 6, max = 100)
+        String repeatPassword,
+
+        @NotNull
+        String firstName,
+        @NotNull
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java:24]: Better use NotBlank with strings

File: src/main/java/com/example/bookstore/dto/user/UserResponseDto.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package com.example.bookstore.dto.user;
+
+public record UserResponseDto(
+        Long id,
+        String email
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/dto/user/UserResponseDto.java:5]: you can return all fields without password

File: src/main/java/com/example/bookstore/model/User.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    private Long id;
+    @Column(nullable = false, unique = true)
+    private String email;
+    @Column(nullable = false)
+    private String password;
+    @Column(nullable = false)
+    private String firstName;
+    @Column(nullable = false)
+    private String lastName;
+    private String shippingAddress;
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/model/User.java:40]:
```suggestion
    private Long id;
    
    @Column(nullable = false, unique = true)
    private String email;
    
    @Column(nullable = false)
    private String password;
    
    @Column(nullable = false)
    private String firstName;
    
    @Column(nullable = false)
    private String lastName;
    private String shippingAddress;
```
Use the same code style for each class

File: src/main/java/com/example/bookstore/model/User.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToMany
+    @JoinTable(name = "user_role",
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/model/User.java:45]:
```suggestion
    @JoinTable(name = "users_roles",

```

File: src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto request)
+            throws RegistrationException {
+        if (userRepository.findByEmail(request.email()).isPresent()) {
+            throw new RegistrationException("User with this email already exists");
+        }
+        User user = userMapper.toUser(request);
+        user.setPassword(passwordEncoder.encode(request.password()));
+        user.setRoles(Collections.singleton(roleRepository.findByName(Role.RoleName.ROLE_USER)));
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/bookstore/service/impl/UserServiceImpl.java:33]:
```suggestion
        user.setRoles(Set.of(roleRepository.findByName(Role.RoleName.ROLE_USER)));

```

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! 👏 Let’s  improve your solution ;)

File: src/main/java/com/example/bookstore/config/SecurityConfig.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Bean
+    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+        return http
+                .cors(AbstractHttpConfigurer::disable)
+                .csrf(AbstractHttpConfigurer::disable)
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers(
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/config/SecurityConfig.java:41]:
```suggestion
                                .requestMatchers("/auth/**", "/error", "/swagger-ui/**")

```

File: src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+
+        @NotBlank
+        @Size(min = 6, max = 100)
+        String password,
+
+        @NotBlank
+        @Size(min = 6, max = 100)
+        String repeatPassword,
+
+        @NotBlank
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/dto/user/UserRegistrationRequestDto.java:22]:
```suggestion
        @NotBlank
        String firstName,
        
        @NotBlank
        String lastName,
        
        String shippingAddress
```

File: src/main/java/com/example/bookstore/security/CustomUserDetailsService.java [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/security/CustomUserDetailsService.java:16]:
```suggestion
    public UserDetails loadUserByUsername(String username) {

```

File: src/main/java/com/example/bookstore/dao/UserRepository.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package com.example.bookstore.dao;
+
+import com.example.bookstore.model.User;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+    Optional<User> existsByEmail(String email);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/bookstore/dao/UserRepository.java:10]:
```suggestion
    boolean existsByEmail(String email);

```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: email
+                  type: varchar(255)
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml:18]: Check all sizes - they should be the same as in requests

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

**Pull request merged**